crates/python/src/qpu/api.rs

```diff
@@ -187,34 +191,83 @@ pub struct ExecutionResults {
     /// QPU execution duration.
     #[pyo3(get)]
     pub execution_duration_microseconds: Option<u64>,
+    #[pyo3(get)]
+    pub memory: HashMap<String, Option<PyMemoryValues>>,
 }
 
 #[pymethods]
 impl ExecutionResults {
     #[new]
     fn new(
         buffers: HashMap<String, ExecutionResult>,
+        memory: HashMap<String, Option<PyMemoryValues>>,
         execution_duration_microseconds: Option<u64>,
     ) -> Self {
         Self {
             buffers,
             execution_duration_microseconds,
+            memory,
         }
     }
 }
 
-impl From<ControllerJobExecutionResult> for ExecutionResults {
-    fn from(value: ControllerJobExecutionResult) -> Self {
-        let buffers = value
+impl ExecutionResults {
+    fn from_controller_job_execution_result(
+        py: Python<'_>,
+        result: ControllerJobExecutionResult,
+    ) -> PyResult<Self> {
+        let buffers = result
             .readout_values
             .into_iter()
             .filter_map(|(key, val)| val.values.map(|values| (key, values.into())))
             .collect();
 
-        Self {
+        let memory = result.memory_values.iter().try_fold(
+            HashMap::with_capacity(result.memory_values.len()),
+            |mut acc, (key, value)| -> PyResult<HashMap<_, _>> {
+                acc.insert(
+                    key.clone(),
+                    match &value.value {
+                        Some(data_value::Value::Binary(value)) => {
+                            Some(PyMemoryValues::from_binary(
+                                py,
+                                value
+                                    .data
+                                    .iter()
+                                    .map(|v| v.to_object(py).extract(py))
+                                    .collect::<PyResult<Vec<_>>>()?,
+                            )?)
+                        }
+                        Some(data_value::Value::Integer(value)) => {
+                            Some(PyMemoryValues::from_integer(
+                                py,
+                                value
+                                    .data
+                                    .iter()
+                                    .map(|v| v.to_object(py).extract(py))
+                                    .collect::<PyResult<Vec<_>>>()?,
+                            )?)
+                        }
+                        Some(data_value::Value::Real(value)) => Some(PyMemoryValues::from_real(
+                            py,
+                            value
+                                .data
+                                .iter()
+                                .map(|v| v.to_object(py).extract(py))
+                                .collect::<PyResult<Vec<_>>>()?,
+                        )?),
+                        None => None,
```

Comment:
Why not just skip this case? It's mostly an artifact of protobuf making everything optional. If the values are missing, is there benefit to not just treating it as if that key wasn't in the map at all?

Reply:
I had the same thought just now, and this thread was marked resolved without any clear indication why.

My suggestion would be a `.filter_map()` returning `Option<PyResult<PyMemoryValues>>` followed by a `.collect()`, instead of the `try_fold`

Reply:
`try_fold` isn't used because of `Option`, but because the conversion to Python types from rust types is fallible and it lets us handle those errors in the iterator.

Reply:
Ah, I hadn't seen this comment since the diff was outdated, but note that a `map` iterator can be fallible if you `collect::<Result<_,_>>()?`

Reply:
Yes but you lose the syntactical sugar of `?` and `try_fold` will end early if it hits an error.